src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/helpers/sync_to_s3.py

```diff
@@ -0,0 +1,820 @@
+        LOGGER.warning(
+            "Input warning: The file_extension filter is not applied "
+            "when you are trying to sync a single file to S3. "
+            "The --extension <extension> argument is only compatible when "
+            "performing a --recursive directory sync."
+        )
+
+
+
+def sync_files(
+    s3_client: any,
+    local_path: str,
+    file_extension: str,
+    s3_url: str,
+    recursive: bool,
+    delete: bool,
+    metadata_key_values: dict[str, str],
+):
+    """
+    Sync files using the S3 client from the local_path, matching the local_glob
+    to the specific s3_url.
+
+    Args:
+        s3_client (Boto3.Client): The Boto3 S3 Client to interact with when
+            a file needs to be deleted.
+        file_extension (str): The extension to search with for files inside a
+            specific path. For example, ".yml" will return all YAML files,
+            including those in sub directories.
+        s3_url (str): The S3 URL to use, for example
+            S3://bucket/specific/prefix.
+    """
+    s3_url_details = urlparse(s3_url)
+    s3_bucket = s3_url_details.netloc
+    s3_prefix = clean_s3_prefix(str(s3_url_details.path))
+
+    ensure_valid_input(
+        local_path,
+        file_extension,
+        s3_url,
+        s3_bucket,
+        s3_prefix,
+        recursive,
+    )
+
+    local_files = get_local_files(local_path, file_extension, recursive)
+
+    s3_objects = get_s3_objects(
+        s3_client,
+        s3_bucket,
+        s3_prefix,
+        file_extension,
+        recursive,
+    )
+
+    upload_changed_files(
+        s3_client,
+        s3_bucket,
+        s3_prefix,
+        local_files,
+        s3_objects,
+        metadata_key_values,
+    )
+    if delete:
+        delete_stale_objects(
+            s3_client,
+            s3_bucket,
+            s3_prefix,
+            local_files,
+            s3_objects,
+        )
+
+
+def main():  # pylint: disable=R0915
+    """Main function to sync files to S3"""
+
+    options = docopt(__doc__, version=ADF_VERSION, options_first=True)
+    # In case the user asked for verbose logging, increase
+    # the log level to debug.
+    if options["--verbose"] > 0:
+        LOGGER.setLevel(logging.DEBUG)
+    if options["--verbose"] > 1:
+        # Also enable DEBUG mode for other libraries, like boto3
+        logging.basicConfig(level=logging.DEBUG)
+
+    LOGGER.debug("Input arguments: %s", options)
+
+    local_path = options.get('SOURCE_PATH')
+    # file_extension = options.get('--extension') or ""
+    file_extension = options.get('--extension') or ""
+    s3_url = options.get('DESTINATION_S3_URL')
+    recursive = options.get('--recursive', False)
+    delete = options.get('--delete', False)
+
+    # Convert metadata key and value lists into a dictionary
+    # metadata_key_values = dict(map(
+    #     lambda idx: (options['<key>'][idx], options['<value>'][idx]),
+    #     range(len(options['<key>'])),
+    # ))
```

Comment:
Is this needed

Reply:
Good one, I replaced that. Will remove it.